test/elixir/test/auth_cache_test.exs

```diff
@@ -0,0 +1,215 @@
+
+      _ ->
+        assert false
+    end
+  end
+
+  defp compact(db_name) do
+    resp = Couch.post("/#{db_name}/_compact")
+    assert resp.status_code == 202
+    resp.body
+  end
+
+  def save_doc(db_name, body) do
+    resp = Couch.put("/#{db_name}/#{body["_id"]}", body: body)
+    assert resp.status_code in [201, 202]
+    assert resp.body["ok"]
+    {:ok, resp}
+  end
+
+  def delete_doc(db_name, body) do
+    resp = Couch.delete("/#{db_name}/#{body["_id"]}", query: [rev: body["_rev"]])
+    assert resp.status_code in [200, 202]
+    assert resp.body["ok"]
+    {:ok, resp}
+  end
+
+  defp test_fun(db_name) do
+    fdmanana =
+      prepare_user_doc([
+        {:name, "fdmanana"},
+        {:password, "qwerty"},
+        {:roles, ["dev"]}
+      ])
+
+    {:ok, resp} = create_doc(db_name, fdmanana)
+    fdmanana = Map.put(fdmanana, "_rev", resp.body["rev"])
+
+    chris =
+      prepare_user_doc([
+        {:name, "chris"},
+        {:password, "the_god_father"},
+        {:roles, ["dev", "mafia", "white_costume"]}
+      ])
+
+    create_doc(db_name, chris)
+
+    joe =
+      prepare_user_doc([
+        {:name, "joe"},
+        {:password, "functional"},
+        {:roles, ["erlnager"]}
+      ])
+
+    create_doc(db_name, joe)
+
+    johndoe =
+      prepare_user_doc([
+        {:name, "johndoe"},
+        {:password, "123456"},
+        {:roles, ["user"]}
+      ])
+
+    create_doc(db_name, johndoe)
+
+    assert_cache(:expect_miss, "fdmanana", "qwerty")
+    assert_cache(:expect_hit, "fdmanana", "qwerty")
+    assert_cache(:expect_miss, "chris", "the_god_father")
+    assert_cache(:expect_miss, "joe", "functional")
+    assert_cache(:expect_miss, "johndoe", "123456")
+
+    # It's a MRU cache, joe was removed from cache to add johndoe
+    # BUGGED assert_cache(:expect_miss, "joe", "functional")
+
+    assert_cache(:expect_hit, "fdmanana", "qwerty")
+
+    fdmanana = Map.replace!(fdmanana, "password", "foobar")
+    {:ok, resp} = save_doc(db_name, fdmanana)
+    second_rev = resp.body["rev"]
+
+    # Cache was refreshed
+    # BUGGED
+    # assert_cache(:expect_hit, "fdmanana", "qwerty", :expect_login_fail)
+    # assert_cache(:expect_hit, "fdmanana", "foobar")
+
+    # and yet another update
+    fdmanana = Map.replace!(fdmanana, "password", "javascript")
+    fdmanana = Map.replace!(fdmanana, "_rev", second_rev)
+    {:ok, resp} = save_doc(db_name, fdmanana)
+    third_rev = resp.body["rev"]
+    fdmanana = Map.replace!(fdmanana, "_rev", third_rev)
+
+    # Cache was refreshed
+    # BUGGED
+    # assert_cache(:expect_hit, "fdmanana", "foobar", :expect_login_fail)
+    # assert_cache(:expect_hit, "fdmanana", "javascript")
+
+    delete_doc(db_name, fdmanana)
+
+    assert_cache(:expect_hit, "fdmanana", "javascript", :expect_login_fail)
+
+    # login, compact authentication DB, login again and verify that
+    # there was a cache hit
+    assert_cache(:expect_hit, "johndoe", "123456")
```

Comment:
**comment for other reviewers**: The correspondent lines in the JavaScript test [are](https://github.com/apache/couchdb/blob/master/test/javascript/tests/auth_cache.js#L248:L249):
```
// XXX BUGGED T(misses_after === (misses_before + 1));
// XXX BUGGED T(hits_after === hits_before);
```
- the check is commented out
- the commented out lines expect a cache **miss**